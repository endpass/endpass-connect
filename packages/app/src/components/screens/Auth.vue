<template>
  <create-account-form
    v-if="needAccount"
    @request="handleAccountRequest"
    @cancel="handleAuthCancel"
  />
  <auth-form
    v-else
    :inited="inited"
    :loading="loading"
    :show-email="!authorized && !sent"
    :message="formMessage"
    :error="error"
    @cancel="handleAuthCancel"
    @submit="handleAuthSubmit"
  />
</template>

<script>
import isEmpty from 'lodash/isEmpty';
import { mapActions, mapState } from 'vuex';
import AuthForm from '../AuthForm.vue';
import CreateAccountForm from '../CreateAccountForm.vue';

export default {
  name: 'Auth',

  data: () => ({
    error: null,
    needAccount: false,
  }),

  computed: {
    ...mapState({
      inited: state => state.core.inited,
      loading: state => state.core.loading,
      sent: state => state.accounts.linkSent,
      accounts: state => state.accounts.accounts,
    }),

    authorized() {
      return !!this.accounts;
    },

    confirmed() {
      return this.authorized && this.sent;
    },

    isAccountsEmpty() {
      return isEmpty(this.accounts);
    },

    formMessage() {
      if (this.authorized) {
        return 'You are currently logged in.';
      }

      if (!this.authorized && this.sent) {
        return 'An email with authorization link was sent on your address. Open it in the same browser to sign in.';
      }

      return 'Log in to your Endpass account to access site actions';
    },
  },

  watch: {
    authorized() {
      const { confirmed, isAccountsEmpty, awaitAccountCreate } = this;

      if (confirmed && isAccountsEmpty) {
        this.needAccount = true;
        awaitAccountCreate();
      } else if (confirmed && !isAccountsEmpty) {
        this.needAccount = false;
      }
    },

    accounts() {
      const { confirmed, isAccountsEmpty, confirmAuth } = this;

      if (confirmed && !isAccountsEmpty) {
        confirmAuth();
      }
    },
  },

  methods: {
    ...mapActions([
      'auth',
      'cancelAuth',
      'confirmAuth',
      'awaitAuthConfirm',
      'awaitAccountCreate',
      'sendReadyMessage',
      'openCreateAccountPage',
    ]),

    async handleAuthSubmit(email) {
      try {
        await this.auth(email);
        await this.awaitAuthConfirm();
      } catch (err) {
        this.handleAuthError(err);
      }
    },

    async handleAccountRequest() {
      this.openCreateAccountPage();
    },

    handleAuthCancel() {
      this.cancelAuth();
    },

    handleWindowClose() {
      this.cancelAuth();
    },

    handleAuthError(error) {
      this.error = error.message || 'Unexpected error, try login later';
    },
  },

  async created() {
    window.addEventListener('beforeunload', this.handleWindowClose);

    await this.sendReadyMessage();

    this.handleAuthStatusChange();
  },

  components: {
    AuthForm,
    CreateAccountForm,
  },
};
</script>
