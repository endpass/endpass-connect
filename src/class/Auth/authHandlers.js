import ConnectError from '@endpass/class/ConnectError';
import { METHODS } from '@/constants';

const { ERRORS } = ConnectError;

const authStatus = auth => status => {
  // eslint-disable-next-line no-param-reassign
  auth.isLogin = status;
};

const logout = auth => async (payload, req) => {
  try {
    const res = await auth.logout();

    req.answer({
      status: res,
    });
  } catch (error) {
    const code = (error && error.code) || ERRORS.AUTH_LOGOUT;
    throw ConnectError.create(code);
  }
};

export default {
  [METHODS.AUTH_STATUS]: authStatus,
  [METHODS.LOGOUT_REQUEST]: logout,
};
