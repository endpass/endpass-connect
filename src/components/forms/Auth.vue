<template>
  <form data-test="auth-form" @submit.prevent="emitSubmit">
    <form-field>
      <message data-test="form-message"
        >Log in to your Endpass account to access site actions</message
      >
    </form-field>
    <form-field v-if="error">
      <message :error="true" data-test="error-message">{{ error }}</message>
    </form-field>
    <form-field>
      <div class="auth__fields-as-line">
        <v-input
          v-model="email"
          :invalid="!isEmailValid"
          :autofocus="true"
          name="email"
          type="email"
          placeholder="Enter your email..."
          data-test="email-input"
        />
        <v-button
          :disabled="!termsAccepted || !isEmailValid || loading"
          :submit="true"
          type="primary"
          data-test="submit-button"
          >{{ primaryButtonLabel }}</v-button
        >
      </div>
    </form-field>
    <form-controls>
      <google-auth-button />
      <git-auth-button />
    </form-controls>
    <form-controls>
      <v-checkbox v-model="termsAccepted">
        I accept the
        <a href="https://endpass.com/terms/" target="_blank"
          >Terms of Service</a
        >
        and
        <a href="https://endpass.com/privacy/" target="_blank">Privacy Policy</a
        >.
      </v-checkbox>
    </form-controls>
  </form>
</template>

<script>
import Vue from 'vue';

import VCheckbox from '@endpass/ui/dist/components/VCheckbox';
import VFrame from '../VFrame.vue';
import VInput from '../VInput.vue';
import VButton from '../VButton.vue';
import GoogleAuthButton from '@/components/GoogleAuthButton.vue';
import GitAuthButton from '@/components/GitAuthButton.vue';
import Message from '../Message.vue';
import FormField from '../FormField.vue';
import FormControls from '../FormControls.vue';

Vue.component(VCheckbox);

export default {
  name: 'AuthForm',

  props: {
    inited: {
      type: Boolean,
      default: false,
    },

    loading: {
      type: Boolean,
      default: false,
    },

    error: {
      type: String,
      default: null,
    },
  },

  data: () => ({
    termsAccepted: true,
    email: '',
  }),

  computed: {
    primaryButtonLabel() {
      return !this.loading ? 'Log in' : 'Loading...';
    },

    isEmailValid() {
      return /[a-zA-Z._\-0-9]+@[a-z0-9]+\.[a-z]{2,}/g.test(this.email);
    },
  },

  methods: {
    emitSubmit() {
      if (this.isEmailValid) {
        this.$emit('submit', this.email);
      }
    },
  },
  components: {
    VCheckbox,
    VFrame,
    VButton,
    VInput,
    GoogleAuthButton,
    GitAuthButton,
    Message,
    FormField,
    FormControls,
  },
};
</script>
<style>
.auth__fields-as-line {
  display: flex;
}
</style>
