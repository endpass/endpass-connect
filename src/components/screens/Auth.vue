<template>
  <v-frame :loading="!inited">
    <create-account-form
      v-if="authorized && isAccountsEmpty"
      :closable="isDialog"
      @request="handleAccountRequest"
      @cancel="handleAuthCancel"
    />
    <otp-form 
      v-else-if="otpEmail" 
      :closable="isDialog"
      @submit="handleOtpSubmit"
      @cancel="handleAuthCancel"
    />
    <message-form 
      v-else-if="!authorized && sent"
      :closable="isDialog"
      message="An email with authorization link was sent on your address. Open it in the same browser to sign in."
      @cancel="handleAuthCancel"
    />
    <auth-form
      v-else
      :inited="inited"
      :loading="loading"
      :error="error"
      :closable="isDialog"
      @cancel="handleAuthCancel"
      @submit="handleAuthSubmit"
    />
  </v-frame> 
</template>

<script>
import isEmpty from 'lodash/isEmpty';
import { mapActions, mapGetters, mapState } from 'vuex';
import VFrame from '../VFrame.vue';
import AuthForm from '../AuthForm.vue';
import OtpForm from '../OtpForm.vue';
import MessageForm from '../MessageForm.vue';
import CreateAccountForm from '../CreateAccountForm.vue';

export default {
  name: 'Auth',

  data: () => ({
    error: null,
    needAccount: false,
  }),

  computed: {
    ...mapState({
      inited: state => state.core.inited,
      loading: state => state.core.loading,
      sent: state => state.accounts.linkSent,
      otpEmail: state => state.accounts.otpEmail,
      accounts: state => state.accounts.accounts,
    }),
    ...mapGetters(['isDialog']),

    authorized() {
      return !!this.accounts;
    },

    confirmed() {
      return this.authorized && this.sent;
    },

    isAccountsEmpty() {
      return isEmpty(this.accounts);
    },
  },

  watch: {
    authorized: {
      handler() {
        this.handleAuthorizationDataChange();
      },
      immediate: true,
    },

    accounts: {
      handler() {
        this.handleAuthorizationDataChange();
      },
      immediate: true,
    },
  },

  methods: {
    ...mapActions([
      'auth',
      'cancelAuth',
      'confirmAuth',
      'confirmAuthViaOtp',
      'awaitAuthMessage',
      'awaitAuthConfirm',
      'awaitAccountCreate',
      'openCreateAccountPage',
    ]),

    async handleOtpSubmit(code) {
      try {
        await this.confirmAuthViaOtp({
          email: this.otpEmail,
          code,
        });
      } catch (err) {
        console.error('handle error', err);
        this.handleAuthError(err);
      }
    },

    async handleAuthSubmit(email) {
      try {
        await this.auth(email);
        await this.awaitAuthConfirm();
      } catch (err) {
        console.error(err);
        this.handleAuthError(err);
      }
    },

    handleAuthorizationDataChange() {
      const {
        authorized,
        awaitAccountCreate,
        isAccountsEmpty,
        confirmAuth,
      } = this;

      if (authorized && isAccountsEmpty) {
        awaitAccountCreate();
      } else if (authorized && !isAccountsEmpty) {
        confirmAuth();
      }
    },

    async handleAccountRequest() {
      this.openCreateAccountPage();
    },

    handleAuthCancel() {
      this.cancelAuth();
    },

    handleWindowClose() {
      this.cancelAuth();
    },

    handleAuthError(error) {
      this.error = error.message || 'Unexpected error, try login later';
    },
  },

  async created() {
    if (this.isDialog) {
      window.addEventListener('beforeunload', this.handleWindowClose);
      this.awaitAuthMessage();
    }
  },

  components: {
    VFrame,
    AuthForm,
    OtpForm,
    MessageForm,
    CreateAccountForm,
  },
};
</script>
