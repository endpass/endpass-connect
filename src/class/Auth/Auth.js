import ConnectError from '@endpass/class/ConnectError';
import { METHODS } from '@/constants';
import HandlersFactory from '@/class/HandlersFactory';
import authHandlers from '@/class/Auth/authHandlers';

const { ERRORS } = ConnectError;

export default class Auth {
  constructor({ dialog, options = {} }) {
    this.dialog = dialog;
    this.isServerLogin = false;
    this.haveDemoData = !!options.demoData;
    this.handleEvent = HandlersFactory.createHandleEvent(this, authHandlers);
  }

  get isLogin() {
    if (this.haveDemoData) {
      return true;
    }

    return this.isServerLogin;
  }

  set isLogin(value) {
    this.isServerLogin = value;
  }

  /**
   * Open application on auth screen and waits result (success of failure)
   * @public
   * @throws {Error} If authentication failed
   * @returns {Promise<boolean>} Auth result, check `status` property to
   *  know about result
   */
  async auth(redirectUrl) {
    const toPath = redirectUrl || window.location.href;

    const res = await this.dialog.ask(METHODS.AUTH, {
      redirectUrl: toPath,
    });

    if (!res.status) {
      throw ConnectError.create(res.code || ERRORS.AUTH);
    }

    this.isLogin = true;

    return {
      payload: res.payload,
      status: res.status,
    };
  }

  /**
   * Send request to logout through injected bridge bypass application dialog
   * @public
   * @throws {Error} If logout failed
   * @returns {Promise<Boolean>}
   */
  async logout() {
    const res = await this.dialog.ask(METHODS.LOGOUT);

    if (!res.status) {
      throw ConnectError.create(res.code || ERRORS.AUTH_LOGOUT);
    }

    this.isLogin = false;

    return res.status;
  }
}
