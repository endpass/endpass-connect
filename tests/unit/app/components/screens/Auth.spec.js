import Vuex from 'vuex';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import Auth from '@@/app/src/components/screens/Auth.vue';

const localVue = createLocalVue();

localVue.use(Vuex);

describe('Auth', () => {
  let store;
  let storeData;
  let wrapper;
  let accountsModule;
  let coreModule;

  beforeEach(() => {
    coreModule = {
      state: {
        inited: true,
        loading: false,
      },
      actions: {
        sendReadyMessage: jest.fn(),
      },
    };
    accountsModule = {
      state: {
        linkSent: false,
        accounts: null,
      },
      actions: {
        auth: jest.fn(),
        cancelAuth: jest.fn(),
        confirmAuth: jest.fn(),
        awaitAuthConfirm: jest.fn(),
        awaitAccountCreate: jest.fn(),
      },
    };
    storeData = {
      modules: {
        accounts: accountsModule,
        core: coreModule,
      },
    };
    store = new Vuex.Store(storeData);
    wrapper = shallowMount(Auth, {
      localVue,
      store,
    });
  });

  describe('render', () => {
    it('should correctly render Auth component', () => {
      expect(wrapper.name()).toBe('Auth');
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('behavior', () => {
    it('should not confirm auth by default on mount', () => {
      expect(accountsModule.actions.confirmAuth).not.toBeCalled();
    });

    it('should not confirm auth if link was sent but authorization status was not changed', () => {
      store.state.accounts.linkSent = true;

      expect(accountsModule.actions.confirmAuth).not.toBeCalled();
    });

    it('should not confirm auth if link was sent and authorization status was changed but account are empty', () => {
      store.state.accounts.linkSent = true;
      store.state.accounts.accounts = [];

      expect(accountsModule.actions.confirmAuth).not.toBeCalled();
    });

    it('should confirm auth if link was sent and authorization status was changed', () => {
      store.state.accounts.linkSent = true;
      store.state.accounts.accounts = ['0x0'];

      expect(accountsModule.actions.confirmAuth).toBeCalled();
    });

    describe('auth form logic', () => {
      it('should request auth on form submit', () => {
        // TODO Have troubles with triggering event from stub, solve it when possivble
        wrapper.vm.handleAuthSubmit('foo@bar.baz');

        expect(accountsModule.actions.auth).toBeCalled();
      });

      it('should cancel auth on form cancel', () => {
        // TODO Have troubles with triggering event from stub, solve it when possivble
        wrapper.vm.handleAuthCancel();

        expect(accountsModule.actions.cancelAuth).toBeCalled();
      });
    });
  });
});
