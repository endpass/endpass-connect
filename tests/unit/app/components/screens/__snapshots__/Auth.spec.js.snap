// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth render should correctly render Auth component 1`] = `
<v-frame-stub>
  <auth-form-stub inited="true" closable="true"></auth-form-stub>
</v-frame-stub>
`;
